apparmor/probe.go

```diff
@@ -0,0 +1,118 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package apparmor
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// FeatureLevel encodes the kind of support for apparmor found on this system.
+type FeatureLevel int
+
+const (
+	// None indicates that apparmor is not enabled.
+	None FeatureLevel = iota
+	// Partial indicates that apparmor is enabled but some features are missing.
+	Partial
+	// Full indicates that all features are supported.
+	Full
+)
+
+var (
+	// featureSysPath points to the sysfs directory where apparmor features are listed.
+	featuresSysPath = "/sys/kernel/security/apparmor/features"
+	// requiredFeatures are the apparmor features needed for strict confinement.
+	requiredFeatures = []string{
+		"caps",
+		"dbus",
+		"domain",
+		"file",
+		"mount",
+		"namespaces",
+		"network",
+		"ptrace",
+		"rlimit",
+		"signal",
+	}
+)
+
+// Probe checks which apparmor features are available.
+//
+// The error is returned whenever less-than-full support is detected.
+func Probe() (FeatureLevel, error) {
+	_, err := os.Stat(featuresSysPath)
```

Comment:
This could be written in a single line: `if _, err := ...; err != nil {`

Reply:
Will do

Reply:
Done